sympy/matrices/matrices.py

```diff
@@ -754,9 +754,16 @@ def num_mat_mul(k1, k2):
     searches for this automatically.
     """
     # Deal with Mul._kind_dispatcher's commutativity
-    elemk = Mul._kind_dispatcher(NumberKind, k2.element_kind)
+    # XXX: this function is called with either k1 or k2 as MatrixKind because
+    # the Mul kind dispatcher is commutative. Maybe it shouldn't be. Need to
+    # swap the args here because NumberKind doesn't have an element_kind
+    # attribute.
+    if not isinstance(k2, MatrixKind):
+        k1, k2 = k2, k1
```

Comment:
@JSS95 how should this be handled?

The purpose of commutative dispatch is (I presume) so that you can avoid writing two handlers. However KindDispatcher registers the same function with the arguments in either order:
https://github.com/sympy/sympy/blob/b4c1da60fa4e8201b2b929083df2610da7a5b52f/sympy/core/kind.py#L288-L292

Reply:
Could fix it with this:
```diff
diff --git a/sympy/core/kind.py b/sympy/core/kind.py
index 43c66ec3d9..16de28549a 100644
--- a/sympy/core/kind.py
+++ b/sympy/core/kind.py
@@ -288,7 +288,8 @@ def register(self, *types, **kwargs):
         def _(func):
             self._dispatcher.add(types, func, **kwargs)
             if self.commutative:
-                self._dispatcher.add(tuple(reversed(types)), func, **kwargs)
+                revfunc = lambda k1, k2: func(k2, k1)
+                self._dispatcher.add(tuple(reversed(types)), revfunc, **kwargs)
         return _
 
     def __call__(self, *args, **kwargs):
```
I'm not sure if it would be better just to handle the commutative case in `dispatch_kinds` though. In that case the fix is just:
```diff
diff --git a/sympy/core/kind.py b/sympy/core/kind.py
index 43c66ec3d9..991394d7cc 100644
--- a/sympy/core/kind.py
+++ b/sympy/core/kind.py
@@ -287,8 +287,7 @@ def register(self, *types, **kwargs):
 
         def _(func):
             self._dispatcher.add(types, func, **kwargs)
-            if self.commutative:
-                self._dispatcher.add(tuple(reversed(types)), func, **kwargs)
+
         return _
 
     def __call__(self, *args, **kwargs):
```

Reply:
Maybe if non-commutative dispatching rules are allowed then it doesn't make sense to have any commutative matchers as well.

Reply:
Also `if isinstance(expr.kind, MatrixKind)` is not a great way to write this. It would be better to have something like `is_matrix(expr)`.

Reply:
Registering inverse function seems good if it does not break anything. Would you elaborate on the second suggestion about `dispatch_kinds`?

Reply:
The point is that commutativity is being handled in two places:

1. The function is initially registered for both `types` and `reversed(types)`.
2. `dispatch_kinds` checks if a handler is registered for `(t1, t2)` and then for `(t2, t1)`.

The problem with 1. is that it means that although the handler is written as
```
@register(_NumberKind, MatrixKind)
def handler(k1, k2):
    ...
```
the handler can be called with k1 and k2 swapped so e.g. k1 is MatrixKind. That means that the handler has to check which argument is which which basically defeats the point of having a single handler registered for either direction of the arguments.

Then if 1. is removed we can depend on 2. for commutative dispatch. However the comment indicates that the matrix dispatch needs to get the order correct if the `element_kind` is not commutative but there's no way to do that if the handler doesn't know what the order was.